### PR TITLE
funding: fix flake in itest caused by persistent fee param changes

### PR DESCRIPTION
**Pull request by guggero**
Partially reverts changes of #7597, in order to fix an itest flake.

### CODE REVIEW

**Review by ellemouton — APPROVED** [review]
Very nice 🔥  i've never seen the `funlen` linter get triggered 😂  I think this is an OK place to put a `nolint` though

File: .golangci.yml [hunk 1]
```diff
@@ -65,7 +65,7 @@ linters-settings:
   funlen:
     # Checks the number of lines in a function.
     # If lower than 0, disable the check.
-    lines: 200
+    lines: 300
```
Comment (yyforyongyu) [inline .golangci.yml:68]: I think if we have a func this long it usually means we need to break it into smaller pieces for maintenance.

Comment (guggero) [reply .golangci.yml]: Okay, I refactored the code to not be too long. More code for you to review though ;-)

**Comment by guggero** [comment]
Looks like I broke something in a different itest. Will take a look on Monday.

**Comment by guggero** [comment]
> Looks like I broke something in a different itest. Will take a look on Monday.

Looks to have been flakes, looks better after a rebase.

File: funding/manager.go [hunk 2]
```diff
@@ -4555,6 +4542,15 @@ func (f *Manager) getInitialForwardingPolicy(
 	copy(chanID, permChanID[:])
 
 	value, err := f.cfg.Wallet.Cfg.Database.GetInitialFwdingPolicy(chanID)
+
+	// The ErrChannelNotFound error is returned if there was no forwarding
+	// policy stored for this channel. In this case, we'll return the
+	// default forwarding policy. This should only happen when upgrading
+	// from an older version of lnd while a channel is still pending.
+	if errors.Is(err, channeldb.ErrChannelNotFound) {
```
Comment (yyforyongyu) [inline funding/manager.go:4531]: I guess this is a leftover from the previous PR, as we still can't tell that, we there's an `ErrChannelNotFound`, we don't it's due to the upgrading, or something wrong in db. If it's the latter, we will return a default policy, which may cause some hard-to-find bugs. In hindsight, we probably should've migrated the db by storing the default policy for these channels. Non-blocking tho as I don't we can do anything here atm.

File: peer/brontide.go [hunk 3]
```diff
@@ -2437,19 +2438,14 @@ out:
...
 			// defaults to the cap on the total value of outstanding HTLCs.
-			//
-			// TODO(guggero): We should instead pass in the current
-			// forwarding policy from the funding manager to avoid
-			// needing us to update the link once the channel is
-			// announced to the network with custom user values.
-			fwdMinHtlc := lnChan.FwdMinHtlc()
-			defaultPolicy := p.cfg.RoutingPolicy
+			localCfg := newChan.LocalChanCfg
+			initPolicy := newChanReq.forwardingPolicy
```
Comment (yyforyongyu) [inline peer/brontide.go:2442]: I don't think we need to use `forwardingPolicy` here. Instead we can do a query using `p.cfg.ChannelDB.GetInitialFwdingPolicy()`.

File: lnpeer/peer.go [hunk 4]
```diff
@@ -25,7 +51,9 @@ type Peer interface {
 
 	// AddNewChannel adds a new channel to the peer. The channel should fail
 	// to be added if the cancel channel is closed.
-	AddNewChannel(channel *channeldb.OpenChannel, cancel <-chan struct{}) error
+	AddNewChannel(channel *channeldb.OpenChannel,
```
Comment (yyforyongyu) [inline lnpeer/peer.go:54]: Don't think this interface method needs to be expanded to care more than the `*channeldb.OpenChannel`. Since we can access db in `brontide`, we can instead query the policy from db.

File: peer/brontide.go [hunk 5]
```diff
@@ -2437,19 +2438,14 @@ out:
 			// values, as they currently are always set to the default values
 			// at initial channel creation. Note that the maximum HTLC value
 			// defaults to the cap on the total value of outstanding HTLCs.
-			//
-			// TODO(guggero): We should instead pass in the current
-			// forwarding policy from the funding manager to avoid
-			// needing us to update the link once the channel is
-			// announced to the network with custom user values.
-			fwdMinHtlc := lnChan.FwdMinHtlc()
-			defaultPolicy := p.cfg.RoutingPolicy
```
Comment (yyforyongyu) [inline peer/brontide.go:2446]: the removal of `RoutingPolicy` made me think should we remove it? It seems that it's only used when we load channels from startup, and we should instead read the routing policy from db? Needs more check here, looks like the `RoutingPolicy` is created in `server.go` when reading bitcoin configs...

**Review by yyforyongyu — COMMENTED** [review]
> Looks to have been flakes, looks better after a rebase

Confirm they are flakes. Need to seriously fix it now as it seems to be the only one🧐 will check it later this week.

File: peer/brontide.go (hunk 3 above)
Comment (guggero) [reply peer/brontide.go:2442]: That feels very hacky to be honest. First, the peer package would need to make some assumptions over when and how the data is stored in the DB (given that `p.cfg.ChannelDB.GetInitialFwdingPolicy()` only returns the raw bytes, which we could probably change though). Passing in the initial policies feels way more explicit and IMO is easier to understand in the larger picture of the channel funding flow.

File: peer/brontide.go (hunk 5 above)
Comment (guggero) [reply peer/brontide.go:2446]: That's a good idea, though it looks like it's still used in `loadActiveChannels`.

File: peer/brontide.go (hunk 3 above)
Comment (yyforyongyu) [reply peer/brontide.go:2442]: > the peer package would need to make some assumptions over when and how the data is stored in the DB

hmmm...what assumptions do we need to make here?

> Passing in the initial policies feels way more explicit and IMO is easier to understand in the larger picture of the channel funding flow.

Guess I'm on the other side, thinking even passing `*channeldb.OpenChannel` needs to be changed by passing a channel ID or any identifier, then let `brontide` query db to get the current channel state.  And `brontide` needs to be slowly refactored to care only about sending/receiving messages, and doesn't need to know the details of the channel.

Comment (guggero) [reply peer/brontide.go:2442]: Okay, I'll take another look at this some time next week.

File: peer/brontide.go (hunk 5 above)
Comment (guggero) [reply peer/brontide.go:2446]: Took a look at this again. We are reading the routing policies from the DB, but unfortunately they are only stored in the graph DB and not in the actual channel DB. And it can happen that there is no outgoing channel policy for some reason (I've seen some issues around that), so being able to fall back to the default policy instead of failing the channel sounds like a good reason to keep it.

**Comment by guggero** [comment]
@yyforyongyu there's still a bug in the unit test that I need to fix, so not ready for full review. But could you please quickly check if this new approach is more in line with what you had in mind?

**Review by yyforyongyu — COMMENTED** [review]
> But could you please quickly check if this new approach is more in line with what you had in mind?

Yes I like it! Thanks🙏

**Comment by guggero** [comment]
@yyforyongyu I found the unit test flake, was just a timing issue. So this PR is now ready for full review.
@ellemouton I'm re-requesting review from you as well, since quite a few things changed in the meantime.

File: funding/manager.go [hunk 6]
```diff
@@ -2313,7 +2310,7 @@ func (f *Manager) handleFundingCreated(peer lnpeer.Peer,
 	// With a permanent channel id established we can save the respective
 	// forwarding policy in the database. In the channel announcement phase
 	// this forwarding policy is retrieved and applied.
-	err = f.saveInitialFwdingPolicy(channelID, &forwardingPolicy)
+	err = f.saveInitialForwardingPolicy(channelID, &forwardingPolicy)
```
Comment (yyforyongyu) [inline funding/manager.go:2313]: I think the funder receives `accept_channel`, `funding_signed` and `channel_ready`, and the fudee receives `open_channel`, `funding_created`, and `channel_ready`.

This means `handleFundingAccept` and `handleFundingSigned` are both called by the funder, which means the policy is saved twice via `saveInitialForwardingPolicy`.

File: funding/manager.go [hunk 7]
```diff
@@ -3240,7 +3217,7 @@ func (f *Manager) annAfterSixConfs(completeChan *channeldb.OpenChannel,
 		// For private channels we do not announce the channel policy
 		// to the network but still need to delete them from the
 		// database.
-		err = f.deleteInitialFwdingPolicy(chanID)
+		err = f.deleteInitialForwardingPolicy(chanID)
```
Comment (yyforyongyu) [inline funding/manager.go:3220]: Maybe we can keep it? Remember someone asked whether it's possible to change a private channel into a public one. Anyway just a thought non-blocking.

File: funding/manager.go [hunk 8]
```diff
@@ -3628,6 +3607,46 @@ func (f *Manager) handleChannelReady(peer lnpeer.Peer,
 		f.barrierMtx.Unlock()
 	}()
 
+	// Before we can add the channel to the peer, we'll need to ensure that
```
Comment (yyforyongyu) [inline funding/manager.go:3610]: This piece seems to be very dedicated to doing one thing and I think it can fit into its own method, also easier for unit testing(future, hopefully...)

File: channeldb/forwarding_policy.go [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+	initialChannelForwardingPolicyBucket = []byte(
+		"initialChannelFwdingPolicy",
+	)
+)
+
+// InitialForwardingPolicy describes the set of initial constraints that a given
+// ChannelLink is to adhere to when forwarding HTLCs directly after it is first
+// created. This is identical to the htlcswitch.ForwardingPolicy struct, but is
+// used to avoid an import cycle.
+type InitialForwardingPolicy struct {
```
Comment (yyforyongyu) [inline channeldb/forwarding_policy.go:21]: Makes me think we should add a `TODO` to replace `htlcswitch.ForwardingPolicy` with this one.

File: channeldb/forwarding_policy.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+	//  * incomingHtlc.timeLock - timeLockDelta = fwdInfo.OutgoingCTLV
+	//
+	//    where fwdInfo is the forwarding information extracted from the
+	//    per-hop payload of the incoming HTLC's onion packet.
+	TimeLockDelta uint32
+}
+
+// SaveInitialForwardingPolicy saves the serialized forwarding policy for the
+// provided permanent channel id to the initialChannelForwardingPolicyBucket.
+func (c *ChannelStateDB) SaveInitialForwardingPolicy(chanID lnwire.ChannelID,
```
Comment (yyforyongyu) [inline channeldb/forwarding_policy.go:52]: This is missing some unit tests, but I think we'd better do it in another PR to limit the scope.

File: peer/brontide.go [hunk 11]
```diff
@@ -2431,25 +2431,29 @@ out:
...
+			// We'll query the channel DB for the new channel's
+			// initial forwarding policies to determine the policy
+			// we start out with.
+			db := p.cfg.ChannelDB
+			initialPolicy, err := db.GetInitialForwardingPolicy(
```
Comment (yyforyongyu) [inline peer/brontide.go:2438]: So for private channel or zero conf channel, since it's deleted this will always give an error?

File: peer/brontide.go [hunk 12]
```diff
@@ -2431,25 +2431,29 @@ out:
...
-			// TODO(guggero): We should instead pass in the current
-			// forwarding policy from the funding manager to avoid
-			// needing us to update the link once the channel is
-			// announced to the network with custom user values.
-			fwdMinHtlc := lnChan.FwdMinHtlc()
-			defaultPolicy := p.cfg.RoutingPolicy
+			// We'll query the channel DB for the new channel's
+			// initial forwarding policies to determine the policy
+			// we start out with.
+			db := p.cfg.ChannelDB
```
Comment (yyforyongyu) [inline peer/brontide.go:2437]: I think in `brontide`, we can eventually replace this call with `FetchChannelEdgesByOutpoint` to get the policy. Atm the channel is prematurely added to `brontide` by our funding manager, and I have a coming PR to fix it. Baiscally we want to only notify `brontide`(`AddNewChannel`) after both side has processed the `channel_ready` message. This way we should be able to call `FetchChannelEdgesByOutpoint` to get the policy.

In the long run, we probably won't need to care about channel policy in `brontide` at all and let `htlcswitch` fetch the policy when creating the link.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Nice refactor🙏 And love how the build looks✅✅✅

Meanwhile I think we should save the policy to disk in `handleFundingSigned` for funder's side and `handleFundingCreated` in fundee's side, and only do them there in case anything goes wrong in the previous steps of the funding flow.

Also we should delete the saved policy when funding flow failed.

I don't think these are introduced in this PR tho, but need to be fixed nevertheless.

**Comment by yyforyongyu** [comment]
A second thought is we can go with this one and create a new PR to fix issues around saving and deleting policies. Btw I think this one needs to be in `0.17` so we can have a clean build for other PRs cc @saubyk

**Comment by saubyk** [comment]
tagged for 0.17.0

File: funding/manager.go (hunk 7 above)
Comment (guggero) [reply funding/manager.go:3220]: I don't think we should. this record is only stored to "survive" the time from the channel funding being initiated by the user/RPC until it's announced to the graph. And as far as I understand the values don't really have any meaning for private channels. So if we keep the data, we'd also need to make sure to clean them up outside of the funding manager.

It's also worth noting here that the values we store here can be overwritten any time after the channel is confirmed by using the `UpdateChannelPolicy`. So if we ever add an RPC to make a private channel public, I would expect us to just add the same policy parameters to that RPC so the user can specify them at the time of making the channel public (since the desired fee rates might have changed since the channel's creation).

File: peer/brontide.go (hunk 11 above)
Comment (guggero) [reply peer/brontide.go:2438]: The data is deleted only after we've announced the channel to the network, which happens after adding it to `brontide`.

File: peer/brontide.go (hunk 12 above)
Comment (guggero) [reply peer/brontide.go:2437]: Hmm, but won't the edge only be available in the channel graph _after_ we've announced it to the network and handed it to the router? Which I think currently happens after we've handed the channel to `brontide`. But I agree, it would be nice if we could do it that way around, I'm just not sure how much additional refactor that would currently mean. I'll look into it.

File: channeldb/forwarding_policy.go (hunk 9 above)
Comment (guggero) [reply channeldb/forwarding_policy.go:21]: That's a good idea, I'll move it there.

File: funding/manager.go (hunk 6 above)
Comment (guggero) [reply funding/manager.go:2301]: I checked this again to make sure we're not actually saving things twice. There are 2 calls to `saveInitialForwardingPolicy`, one in `handleFundingSigned` (funder) and one in `handleFundingCreated` (fundee). And then a third one in `ensureInitialForwardingPolicy` which is called by `handleChannelReady` (funder and fundee) when making sure we have a forwarding policy before sending the channel to `brontide`.
So I think everything checks out.

File: peer/brontide.go (hunk 12 above)
Comment (guggero) [reply peer/brontide.go]: Discussed this offline, we'll eventually want to refactor this, but not in this PR.

**Comment by guggero** [comment]
@yyforyongyu I took another look and given our offline discussion recently I think this is ready for final review.

File: peer/brontide.go [hunk 13]
```diff
@@ -3737,26 +3738,18 @@ func (p *Brontide) addActiveChannel(c *channeldb.OpenChannel) error {
...
-	defaultPolicy := p.cfg.RoutingPolicy
-	forwardingPolicy := &htlcswitch.ForwardingPolicy{
-		MinHTLCOut:    fwdMinHtlc,
-		MaxHTLC:       c.LocalChanCfg.MaxPendingAmount,
-		BaseFee:       defaultPolicy.BaseFee,
-		FeeRate:       defaultPolicy.FeeRate,
-		TimeLockDelta: defaultPolicy.TimeLockDelta,
+	// We'll query the channel DB for the new channel's initial forwarding
+	// policies to determine the policy we start out with.
+	initialPolicy, err := p.cfg.ChannelDB.GetInitialForwardingPolicy(chanID)
```
Comment (yyforyongyu) [inline peer/brontide.go:3743]: hmmm got the following errors from the neutrino build,
```
2023-07-24 14:48:33.318 [ERR] PEER: Peer(02d733449d425c564e1382347b67c0adf1a9a2c80f6ea68065d9a0271984f8eecd): unable to query for initial forwarding policy: channel not found
2023-07-24 14:48:33.318 [ERR] FNDG: Unable to add new channel 79e0a85a00632862fb0327cb0cbc5c2f125c8cdb995cd907590b0b850776ee27:0 with peer 02d733449d425c564e1382347b67c0adf1a9a2c80f6ea68065d9a0271984f8eecd: unable to query for initial forwarding policy: channel not found
```
which is very weird because I think we only call `AddNewChannel` in funding manager after the policy is saved to db.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Thank you for the latest push! I really like the simplified interaction flow between `brontide` and `link`. However it seems new flakes have been introduced which says the channel cannot be found when querying the policy when we add the link in `brontide`🤓

File: funding/manager.go (hunk 6 above)
Comment (guggero) [reply funding/manager.go:2301]: Correct, that is just for the cases where `lnd` is upgraded between opening a channel and it being confirmed.

File: channeldb/models/channel.go [hunk 14]
```diff
@@ -90,3 +90,41 @@ func (k *CircuitKey) Decode(r io.Reader) error {
...
+
+// ForwardingPolicy describes the set of constraints that a given ChannelLink
+// is to adhere to when forwarding HTLC's. For each incoming HTLC, this set of
+// constraints will be consulted in order to ensure that adequate fees are
+// paid, and our time-lock parameters are respected. In the event that an
+// incoming HTLC violates any of these constraints, it is to be _rejected_ with
+// the error possibly carrying along a ChannelUpdate message that includes the
+// latest policy.
+type ForwardingPolicy struct {
+	// MinHTLC is the smallest HTLC that is to be forwarded.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply channeldb/models/channel.go:102]: hmmm not fixed👀

**Review by yyforyongyu — COMMENTED** [review]
Can we rebase to run the test again? I think the current failure might be fixed by #7518

File: channeldb/models/channel.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (guggero) [reply channeldb/models/channel.go]: Yeah, I fixed it locally but didn't push yet because I wanted to reproduce the flake you mentioned below. But I did rebase and fix this now.

**Comment by guggero** [comment]
Rebased. Wasn't able to reproduce the Neutrino flake on my machine. Let's see how it looks now.

**Review by yyforyongyu — APPROVED** [review]
LGTM🤓

**Review by guggero — COMMENTED** [review]
I can't approve my own PR, but I'm leaving an explicit "LGTM :tada:" here since @yyforyongyu added some commits to it (thank you very much and great find with the barrier!).

**Comment by guggero** [comment]
@ellemouton this is ready for a final review, thanks!

File: funding/manager.go [hunk 15]
```diff
@@ -639,7 +639,7 @@ func (c channelOpeningState) String() string {
 	case markedOpen:
 		return "markedOpen"
 	case channelReadySent:
-		return "channelReady"
+		return "channelReadySent"
```
Comment (ellemouton) [inline funding/manager.go:642]: 👍

File: funding/manager.go [hunk 16]
```diff
@@ -3070,14 +3070,27 @@ func (f *Manager) receivedChannelReady(node *btcec.PublicKey,
 		return false, err
 	}
```
Comment (ellemouton) [inline funding/manager.go:3072]: really helpful commit message - thanks! 🙏

**Review by ellemouton — APPROVED** [review]
Very nice fix 🔥 LGTM!
